manpages/dmtcp_command.1
'\" t
.\" Manual page created with latex2man on Tue Nov 14 06:26:54 EST 2017
.\" NOTE: This file is generated, DO NOT EDIT.
.de Vb
.ft CW
.nf
..
.de Ve
.ft R

.fi
..
.TH "DMTCP\\_COMMAND" "1" "November, 2017" "Distributed MultiThreaded CheckPointing " "Distributed MultiThreaded CheckPointing "
.SH NAME

\fBdmtcp_command\fP
\-\- send a command to the DMTCP coordinator remotely. 
.PP
.SH SYNOPSIS

.PP
\fBdmtcp_commmand\fP
[\fIOPTIONS\fP]
\fIcommand\fP
[\fIcommand...\fP]
.PP
.SH DESCRIPTION

.PP
\fBdmtcp_command\fP
is a tool to send user commands to the 
DMTCP coordinator remotely. 
.PP
.SH OPTIONS

.PP
.TP
\fB\-h\fP, \fB\-\-coord\-host\fP \fIhostname\fP (environment variable DMTCP_COORD_HOST)
 Hostname where \fBdmtcp_coordinator\fP
is run (default: localhost) 
.PP
.TP
\fB\-p\fP, \fB\-\-port\fP \fIport\fP (environment variable DMTCP_PORT)
 Port where \fBdmtcp_coordinator\fP
is run (default: 7779) 
.PP
.TP
\fB\-\-quiet\fP
 Skip copyright notice 
.TP
\fB\-\-help\fP
 Print this message and exit. 
.TP
\fB\-\-version\fP
 Print version information and exit. 
.PP
.SS Commands for Coordinator
.TP
\fB\-s\fP \fB\-\-status\fP
 Print status message 
.TP
\fB\-c\fP, \fB\-\-checkpoint\fP
 Checkpoint all nodes 
.TP
\fB\-bc\fP, \fB\-\-bcheckpoint\fP
 Checkpoint all nodes, blocking until done 
.TP
\fB\-i\fP, \fB\-\-interval\fP \fI<val>\fP
 Update ckpt interval to <val> seconds (0=never) 
.TP
\fB\-k\fP, \fB\-\-kill\fP
 Kill all nodes 
.TP
\fB\-q\fP, \fB\-\-quit\fP
 Kill all nodes and quit 
.PP
.SH REPORTING BUGS

Report bugs to: dmtcp\-forum@lists.sourceforge.net
.br
DMTCP home page: <http://dmtcp.sourceforge.net> 
.PP
.SH SEE ALSO

\fBdmtcp\fP(1),
\fBdmtcp_coordinator\fP(1),
\fBdmtcp_launch\fP(1),
\fBdmtcp_restart\fP(1),
\fBdmtcp_command\fP(1)
.PP
.SH AUTHOR

See /usr/share/doc/dmtcp\-3.0.0/AUTHORS.
.PP
.SH VERSION

.PP
DMTCP version 3.0.0 of November, 2017\&.
.PP
.SH COPYRIGHT

See /usr/share/doc/dmtcp\-3.0.0/COPYING file.
.PP
License LGPLv3+: GNU LGPL version 3 or later <http://gnu.org/licenses/lgpl.html>. 
.PP
This program comes with ABSOLUTELY NO WARRANTY. 
This is free software, and you are welcome to redistribute it under certain 
conditions; see COPYING file for details. 
.PP
.\" NOTE: This file is generated, DO NOT EDIT.


NEWS
DMTCP NEWS -- history of user-visible changes.

Please send DMTCP bug reports via <dmtcp-forum@lists.sourceforge.net>

Version 2.6.0 release notes
===========================
Newer flags for configure:
  * Rename --enable-debug to --enable-logging
  * Add --enable-debug: "-Wall -g3 -O0"  (for debugging DMTCP)
Newer flags for dmtcp_restart:
  * Add --debug-restart-pause flag to dmtcp_restart
Bug fixes and enhancements:
* Fixes for glibc versions greater than or equal to 2.24
* Fix deadlock in system() wrapper when the child crashes
* Fix deadlock when a process is forked in the resume phase (issue #691)
* jsocket: Warn user if peer closes socket while draining (issue #701)
* Fix epoll1 test (initialize addrlen for accept()) (#705)
* Fix to correctly calculate Coordinator/Host IP:
    Affects some distributed applications
* Allow restored stack to grow if needed.
* Fix bug in POSIX timer: race condition manifested in test/timer.c/Ubuntu-18.04
* Modified InfiniBand plugin for more robust support
    (primarily of interest for MPI)
* The floating point environment (fegetenv()) is now restored on restart.
    (Formerly, only the rounding mode (fegetround()) was restored.)
* The current resource limits (rlim_cur) for RLIMIT_NOFILE and RLIMIT_STACK
    are restored if possible.
* Mutex ownership and robust mutexes are now supported if DMTCP is configured
    with --enable-mutex-wrappers.  (However, this configuration can also
    add runtime overhead if mutex operations are called very frequently.)
    [Thanks to Johannes Stoelp, Laurent Buchard, Pankaj Mehta of Synopsys, Inc.]
* Fix bug if stack grows a lot after a restart.
* Improved support for pty's
* util/gdbinit-example added for those who wish to debug DMTCP internals.
* Many bug fixes

Version 2.5.2 release notes
===========================
* All fixes in Release DMTCP-2.4.9 are incorporated in this release.
* An incompatibility of DMTCP with Open MPI 1.10 when using orterun (mpirun)
  was discovered.  This does not affect recent versions, such as Open MPI 2.x.
* In some rare cases, open files were not properly restored due to
  a use-after-free bug.  This is now fixed.
* In some rare cases, one process had created a SysV shared memory object,
  and a different process was assigned to restore it on restart.  This
  was not handled correctly, and is now fixed.
* Correctly restore CPU affinities of threads
* Virtualized SysV shared memory keys to avoid race condition on restart
* Fixed logic for checking if relative path to file was a duplicate
  of another existing path
* The NSCD area for name service caching daemon was not handled correctly
  in CentOS 6.8 and later correctly.  Fixed now.
* The Linux sched.h include file for scheduling of cores was added to
  satisfy some older Linux distros that needed it for compiling DMTCP.
* Fixed a regression in which --enable-debug (for verbose debug logs)
  was not being properly written.
* The DMTCP coordinator was displaying a spurious warning, "Failed to find
  coordinator IP address", because it did not check for a canonical hostname.
  A related issue prevented DMTCP from working properly on some
  SUSE/openSUSE distros.

Version 2.5.1 release notes
===========================
This release mostly provides added robustness.  Two notable items of
added functionality are:
 i. DMTCP_RESTART_PAUSE and DMTCP_RESTART_PAUSE0 environment variables
    for easier debugging upon initial restart
 ii. The --debug-logs flag was added to dmtcp_launch/dmtcp_restart.
     One can now turn on logging individually for separate plugins,
     instead of only turning it on globally.

An incompatibility of DMTCP with Open MPI 1.10 when using orterun (mpirun)
was discovered.  This may also affect some other versions of Open MPI 1.10.
This bug will be fixed in a future release.

* Fixed an issue when starting multiple DMTCP coordinators on same host
  at approximately the same time
* Fixed issue with PBS scheduler for HPC
* Fixed issue when restarting on a different host with a larger
  limit on the number of open file descriptors
* dmtcp_launch/dmtcp_restart now accept '--debug-logs' flag to specify
  which DMTCP plugins should produce logging information
  (It used to be all or nothing.)
* Improved robustness for IB (InfiniBand) plugin
* Fixed DMTCP_RESTART_PAUSE and DMTCP_RESTART_PAUSE0 environment variables
   for debugging upon restart
* The brk() call was failing on restart on Debian due to overly strict assert
* dmtcp_launch was hanging on some RHEL5 and RHEL6 due to deadlock with
    libc low-level locks.  Fixed now.
* Updated tls_pid_offset in DMTCP to handle newer GLIBc (versions > 2.24)
* Fixed launch of 32-bit binary when forking/execing from a 64-bit executable
* Fixed issue that can affect a parent holding a malloc-lock while forking
* Fixed issue when a user thread calls 'dmtcp_get_coord_ckpt_dir()'

Version 2.5.0 release notes
===========================
This release includes a few new plugins and several bug fixes for robustness.
Some of the highlights include:

* Support for InfiniBand UD (in addition to the more common InfiniBand RC).
* Added support for CMA (Cross-Memory Attach):
  `process_vm_readv` and `process_vm_writev`
* Improved multi-arch (mixed 32-/64- bit) support.
* Re-added `--enable-fast-restart`.
* Added a new commandline option `--with-plugin-32` for dmtcp_launch to specify
  32-bit plugins in a 64-bit environment.
* Added `--enable-pthread-mutex-wrappers` configure flag to enable
  `pthread_mutex_{lock,unlock}` wrappers needed for Open MPI.
* Added ability to specify environment file used in the `modify-env` plugin.
* Allow `dmtcp_restart` to be invoked by root. 
* The following new plugins were added:
  * `pathvirt`: to virtualize filesystem paths.
  * `delayresume`: for finer-grained control over resuming of user threads
     during resume/restart.

See the following url for more details:
  https://github.com/dmtcp/dmtcp/pulls?q=is%3Apr+milestone%3A2.5.0+is%3Aclosed


src/dmtcp_command.cpp
/****************************************************************************
 *   Copyright (C) 2006-2010 by Jason Ansel, Kapil Arya, and Gene Cooperman *
 *   jansel@csail.mit.edu, kapil@ccs.neu.edu, gene@ccs.neu.edu              *
 *                                                                          *
 *  This file is part of DMTCP.                                             *
 *                                                                          *
 *  DMTCP is free software: you can redistribute it and/or                  *
 *  modify it under the terms of the GNU Lesser General Public License as   *
 *  published by the Free Software Foundation, either version 3 of the      *
 *  License, or (at your option) any later version.                         *
 *                                                                          *
 *  DMTCP is distributed in the hope that it will be useful,                *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *  GNU Lesser General Public License for more details.                     *
 *                                                                          *
 *  You should have received a copy of the GNU Lesser General Public        *
 *  License along with DMTCP:dmtcp/src.  If not, see                        *
 *  <http://www.gnu.org/licenses/>.                                         *
 ****************************************************************************/

#include <stdio.h>

#include "coordinatorapi.h"
#include "util.h"

#define BINARY_NAME "dmtcp_command"

using namespace dmtcp;

// gcc-4.3.4 -Wformat=2 issues false positives for warnings unless the format
// string has at least one format specifier with corresponding format argument.
// Ubuntu 9.01 uses -Wformat=2 by default.
static const char *theUsage =
  "Usage:  dmtcp_command [OPTIONS] COMMAND [COMMAND...]\n"
  "Send a command to the dmtcp_coordinator remotely.\n\n"
  "Options:\n\n"
  "  -h, --coord-host HOSTNAME (environment variable DMTCP_COORD_HOST)\n"
  "              Hostname where dmtcp_coordinator is run (default: localhost)\n"
  "  -p, --coord-port PORT_NUM (environment variable DMTCP_COORD_PORT)\n"
  "              Port where dmtcp_coordinator is run (default: "
                                                  STRINGIFY(DEFAULT_PORT) ")\n"
  "  --help\n"
  "              Print this message and exit.\n"
  "  --version\n"
  "              Print version information and exit.\n"
  "\n"
  "Commands for Coordinator:\n"
  "    -s, --status:          Print status message\n"
  "    -l, --list:            List connected clients\n"
  "    -c, --checkpoint:      Checkpoint all nodes\n"
// Could add -B as synonym for -bc
  "    -bc, --bcheckpoint:    Checkpoint all nodes, dmtcp_command blocks until"
                                                                      " done\n"
// Could add -K as synonym for -kc
  "    -kc, --kcheckpoint     Checkpoint all nodes, kill all nodes when done\n"
// "    -xc, --xcheckpoint  deprecated synonym for '-kc': kill nodes if done\n"
  "    -i, --interval <val>   Update ckpt interval to <val> seconds (0=never)\n"
  "    -k, --kill             Kill all nodes\n"
  "    -q, --quit             Kill all nodes and quit\n"
  "\n"
  HELP_AND_CONTACT_INFO
  "\n";


// shift args
#define shift argc--, argv++

int
main(int argc, char **argv)
{
  string interval = "";
  string request = "h";

  initializeJalib();

  // No need to initialize the log file.
  // Util::initializeLogFile();

  // process args
  shift;
  while (argc > 0) {
    string s = argv[0];
    if ((s == "--help" || s == "-h") && argc == 1) {
      printf("%s", theUsage);
      return 1;
    } else if ((s == "--version") && argc == 1) {
      printf("%s", DMTCP_VERSION_AND_COPYRIGHT_INFO);
      return 1;
    } else if (argc > 1 &&
               (s == "-h" || s == "--coord-host" || s == "--host")) {
      setenv(ENV_VAR_NAME_HOST, argv[1], 1);
      shift; shift;
    } else if (argc > 1 &&
               (s == "-p" || s == "--coord-port" || s == "--port")) {
      setenv(ENV_VAR_NAME_PORT, argv[1], 1);
      shift; shift;
    } else if (argv[0][0] == '-' && argv[0][1] == 'p' &&
               isdigit(argv[0][2])) { // else if -p0, for example
      setenv(ENV_VAR_NAME_PORT, argv[0] + 2, 1);
      shift;
    } else if (s == "h" || s == "-h" || s == "--help" || s == "?") {
      fprintf(stderr, theUsage, "");
      return 1;
    } else { // else it's a request
      char *cmd = argv[0];

      // ignore leading dashes
      while (*cmd == '-') {
        cmd++;
      }
      if (*cmd == 'k' && *(cmd+1) == 'c') { // if this is "-kc":
        *cmd = 'K';  // Need to disambiguate '-k' from '-kc' (now '-Kc')
      }
      s = cmd;

      if ((*cmd == 'b' || *cmd == 'K') && *(cmd + 1) != 'c') {
        // If blocking ckpt, next letter must be 'c'; else print the usage
        fprintf(stderr, theUsage, "");
        return 1;
      } else if (*cmd == 's' || *cmd == 'i' || *cmd == 'c' || *cmd == 'b' ||
                 *cmd == 'K' || *cmd == 'k' ||
                 *cmd == 'q' || *cmd == 'l') {
        request = s;
        if (*cmd == 'i') {
          if (isdigit(cmd[1])) { // if -i5, for example
            interval = cmd + 1;
          } else { // else -i 5
            if (argc == 1) {
              fprintf(stderr, theUsage, "");
              return 1;
            }
            interval = argv[1];
            shift;
          }
        }
        shift;
      } else {
        fprintf(stderr, theUsage, "");
        return 1;
      }
    }
  }

  int coordCmdStatus = CoordCmdStatus::NOERROR;
  int numPeers;
  int isRunning;
  int ckptInterval;
  char *workerList = NULL;
  char *cmd = (char *)request.c_str();
  switch (*cmd) {
  case 'h':
    fprintf(stderr, theUsage, "");
    return 1;

  case 'i':
    setenv(ENV_VAR_CKPT_INTR, interval.c_str(), 1);
    CoordinatorAPI::connectAndSendUserCommand(*cmd, &coordCmdStatus);
    printf("Interval changed to %s\n", interval.c_str());
    break;
  case 'b':
  case 'K':

    // blocking prefix
    CoordinatorAPI::connectAndSendUserCommand(*cmd, &coordCmdStatus);

    // actual command
    CoordinatorAPI::connectAndSendUserCommand(*(cmd + 1), &coordCmdStatus);
    break;
  case 's':
    CoordinatorAPI::connectAndSendUserCommand(*cmd,
                                              &coordCmdStatus,
                                              &numPeers,
                                              &isRunning,
                                              &ckptInterval);
    break;
  case 'l':
    workerList =
      CoordinatorAPI::connectAndSendUserCommand(*cmd, &coordCmdStatus);
    break;
  case 'c':
  case 'k':
  case 'q':
    workerList =
      CoordinatorAPI::connectAndSendUserCommand(*cmd, &coordCmdStatus);
    break;
  default:
    fprintf(stderr, theUsage, "");
    break;
  }

  // check for error
  if (coordCmdStatus != CoordCmdStatus::NOERROR) {
    switch (coordCmdStatus) {
    case CoordCmdStatus::ERROR_COORDINATOR_NOT_FOUND:
      if (getenv("DMTCP_COORD_PORT") || getenv("DMTCP_PORT")) {
        fprintf(stderr, "Coordinator not found. Please check port and host.\n");
      } else {
        fprintf(stderr,
                "Coordinator not found. "
                "Try specifying port with \'--port\'.\n");
      }
      break;
    case CoordCmdStatus::ERROR_INVALID_COMMAND:
      fprintf(stderr,
              "Unknown command: %c, try 'dmtcp_command --help'\n", *cmd);
      break;
    case CoordCmdStatus::ERROR_NOT_RUNNING_STATE:
      fprintf(stderr,
              "Error, computation not in running state."
              "  Either a checkpoint is\n"
              " currently happening or there are no connected processes.\n");
      break;
    default:
      fprintf(stderr, "Unknown error\n");
      break;
    }
    return 2;
  }

  if(*cmd == 's'){
    printf("Coordinator:\n");
    char *host = getenv(ENV_VAR_NAME_HOST);
    if (host == NULL) {
      host = getenv("DMTCP_HOST");                 // deprecated
    }
    printf("  Host: %s\n", (host ? host : "localhost"));
    char *port = getenv(ENV_VAR_NAME_PORT);
    if (port == NULL) {
      port = getenv("DMTCP_PORT");                 // deprecated
    }
    printf("  Port: %s\n",
           (port ? port : STRINGIFY(DEFAULT_PORT) " (default port)"));
    if (*cmd == 's') {
      printf("Status...\n");
      printf("  NUM_PEERS=%d\n", numPeers);
      printf("  RUNNING=%s\n", (isRunning ? "yes" : "no"));
      if (ckptInterval) {
        printf("  CKPT_INTERVAL=%d\n", ckptInterval);
      } else {
        printf("  CKPT_INTERVAL=0 (checkpoint manually)\n");
      }
    } else {
      if (workerList) {
        printf("%s", workerList);
        JALLOC_HELPER_FREE(workerList);
      }
    }
  }

  return 0;
}
